import sys
from os.path import dirname, realpath, join
sys.path.append('C:/Users/User/FIX/models')
from models import model
from PyQt5 import QtCore, QtGui, QtWidgets
from PyQt5.QtWidgets import QApplication, QWidget, QFileDialog, QTableWidget, QTableWidgetItem
from PyQt5.uic import loadUiType
import pandas as pd
import re

class Ui_Form(object):
    def setupUi(self, Form):
        Form.setObjectName("Form")
        Form.resize(1185, 746)
        Form.setAutoFillBackground(True)
        self.btn_search = QtWidgets.QPushButton(Form)
        self.btn_search.setGeometry(QtCore.QRect(200, 100, 93, 31))
        self.btn_search.setObjectName("btn_search")
        self.tableWidget = QtWidgets.QTableWidget(Form)
        self.tableWidget.setGeometry(QtCore.QRect(70, 140, 1051, 261))
        self.tableWidget.setObjectName("tableWidget")
        self.tableWidget.setColumnCount(0)
        self.tableWidget.setRowCount(0)
        self.lineEdit = QtWidgets.QLineEdit(Form)
        self.lineEdit.setGeometry(QtCore.QRect(20, 20, 1141, 51))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setPointSize(9)
        font.setBold(True)
        font.setWeight(75)
        self.lineEdit.setFont(font)
        self.lineEdit.setObjectName("lineEdit")
        self.label_inputdata = QtWidgets.QLabel(Form)
        self.label_inputdata.setGeometry(QtCore.QRect(70, 110, 121, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(True)
        font.setWeight(75)
        self.label_inputdata.setFont(font)
        self.label_inputdata.setObjectName("label_inputdata")
        self.btn_clasify = QtWidgets.QPushButton(Form)
        self.btn_clasify.setGeometry(QtCore.QRect(340, 470, 93, 31))
        font = QtGui.QFont()
        font.setBold(True)
        font.setWeight(75)
        self.btn_clasify.setFont(font)
        self.btn_clasify.setObjectName("btn_clasify")
        self.frame = QtWidgets.QFrame(Form)
        self.frame.setGeometry(QtCore.QRect(70, 520, 331, 211))
        self.frame.setFrameShape(QtWidgets.QFrame.StyledPanel)
        self.frame.setFrameShadow(QtWidgets.QFrame.Raised)
        self.frame.setObjectName("frame")
        self.label_cf = QtWidgets.QLabel(self.frame)
        self.label_cf.setGeometry(QtCore.QRect(110, 10, 131, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(True)
        font.setWeight(75)
        self.label_cf.setFont(font)
        self.label_cf.setObjectName("label_cf")
        self.label_TP = QtWidgets.QLabel(self.frame)
        self.label_TP.setGeometry(QtCore.QRect(30, 60, 101, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(False)
        font.setWeight(50)
        self.label_TP.setFont(font)
        self.label_TP.setObjectName("label_TP")
        self.label_FP = QtWidgets.QLabel(self.frame)
        self.label_FP.setGeometry(QtCore.QRect(180, 60, 101, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(False)
        font.setWeight(50)
        self.label_FP.setFont(font)
        self.label_FP.setObjectName("label_FP")
        self.label_FN = QtWidgets.QLabel(self.frame)
        self.label_FN.setGeometry(QtCore.QRect(30, 130, 101, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(False)
        font.setWeight(50)
        self.label_FN.setFont(font)
        self.label_FN.setObjectName("label_FN")
        self.label_TN = QtWidgets.QLabel(self.frame)
        self.label_TN.setGeometry(QtCore.QRect(180, 130, 101, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(False)
        font.setWeight(50)
        self.label_TN.setFont(font)
        self.label_TN.setObjectName("label_TN")
        self.frame_2 = QtWidgets.QFrame(self.frame)
        self.frame_2.setGeometry(QtCore.QRect(0, 0, 331, 211))
        self.frame_2.setFrameShape(QtWidgets.QFrame.StyledPanel)
        self.frame_2.setFrameShadow(QtWidgets.QFrame.Raised)
        self.frame_2.setObjectName("frame_2")
        self.label_cf_2 = QtWidgets.QLabel(self.frame_2)
        self.label_cf_2.setGeometry(QtCore.QRect(110, 10, 131, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(True)
        font.setWeight(75)
        self.label_cf_2.setFont(font)
        self.label_cf_2.setObjectName("label_cf_2")
        self.label_TP_2 = QtWidgets.QLabel(self.frame_2)
        self.label_TP_2.setGeometry(QtCore.QRect(30, 60, 101, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(False)
        font.setWeight(50)
        self.label_TP_2.setFont(font)
        self.label_TP_2.setObjectName("label_TP_2")
        self.label_FP_2 = QtWidgets.QLabel(self.frame_2)
        self.label_FP_2.setGeometry(QtCore.QRect(180, 60, 101, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(False)
        font.setWeight(50)
        self.label_FP_2.setFont(font)
        self.label_FP_2.setObjectName("label_FP_2")
        self.label_FN_2 = QtWidgets.QLabel(self.frame_2)
        self.label_FN_2.setGeometry(QtCore.QRect(30, 130, 101, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(False)
        font.setWeight(50)
        self.label_FN_2.setFont(font)
        self.label_FN_2.setObjectName("label_FN_2")
        self.label_TN_2 = QtWidgets.QLabel(self.frame_2)
        self.label_TN_2.setGeometry(QtCore.QRect(180, 130, 101, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(False)
        font.setWeight(50)
        self.label_TN_2.setFont(font)
        self.label_TN_2.setObjectName("label_TN_2")
        self.lineEdit_TP = QtWidgets.QLineEdit(self.frame_2)
        self.lineEdit_TP.setGeometry(QtCore.QRect(30, 80, 81, 22))
        self.lineEdit_TP.setObjectName("lineEdit_TP")
        self.lineEdit_FP = QtWidgets.QLineEdit(self.frame_2)
        self.lineEdit_FP.setGeometry(QtCore.QRect(180, 80, 81, 22))
        self.lineEdit_FP.setObjectName("lineEdit_FP")
        self.lineEdit_FN = QtWidgets.QLineEdit(self.frame_2)
        self.lineEdit_FN.setGeometry(QtCore.QRect(30, 150, 81, 22))
        self.lineEdit_FN.setObjectName("lineEdit_FN")
        self.lineEdit_TN = QtWidgets.QLineEdit(self.frame_2)
        self.lineEdit_TN.setGeometry(QtCore.QRect(180, 150, 81, 22))
        self.lineEdit_TN.setObjectName("lineEdit_TN")
        self.frame_3 = QtWidgets.QFrame(Form)
        self.frame_3.setGeometry(QtCore.QRect(430, 520, 291, 211))
        self.frame_3.setFrameShape(QtWidgets.QFrame.StyledPanel)
        self.frame_3.setFrameShadow(QtWidgets.QFrame.Raised)
        self.frame_3.setObjectName("frame_3")
        self.label_klasifikasi = QtWidgets.QLabel(self.frame_3)
        self.label_klasifikasi.setGeometry(QtCore.QRect(80, 10, 131, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(True)
        font.setWeight(75)
        self.label_klasifikasi.setFont(font)
        self.label_klasifikasi.setObjectName("label_klasifikasi")
        self.label_akurasi = QtWidgets.QLabel(self.frame_3)
        self.label_akurasi.setGeometry(QtCore.QRect(20, 60, 61, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(False)
        font.setWeight(50)
        self.label_akurasi.setFont(font)
        self.label_akurasi.setObjectName("label_akurasi")
        self.lineEdit_akurasi = QtWidgets.QLineEdit(self.frame_3)
        self.lineEdit_akurasi.setGeometry(QtCore.QRect(20, 80, 81, 22))
        self.lineEdit_akurasi.setObjectName("lineEdit_akurasi")
        self.label_inputdata_2 = QtWidgets.QLabel(Form)
        self.label_inputdata_2.setGeometry(QtCore.QRect(80, 470, 161, 16))
        font = QtGui.QFont()
        font.setFamily("Montserrat")
        font.setBold(True)
        font.setWeight(75)
        self.label_inputdata_2.setFont(font)
        self.label_inputdata_2.setObjectName("label_inputdata_2")
        self.lineEdit_jumlahpohon = QtWidgets.QLineEdit(Form)
        self.lineEdit_jumlahpohon.setGeometry(QtCore.QRect(240, 470, 81, 22))
        self.lineEdit_jumlahpohon.setObjectName("lineEdit_jumlahpohon")
        self.checkBox = QtWidgets.QCheckBox(Form)
        self.checkBox.setGeometry(QtCore.QRect(950, 410, 171, 20))
        self.checkBox.setObjectName("checkBox")
        self.btn_open = QtWidgets.QPushButton(Form)
        self.btn_open.setGeometry(QtCore.QRect(950, 440, 93, 31))
        self.btn_open.setObjectName("btn_open")

        self.retranslateUi(Form)
        QtCore.QMetaObject.connectSlotsByName(Form)

    def retranslateUi(self, Form):
        _translate = QtCore.QCoreApplication.translate
        Form.setWindowTitle(_translate("Form", "Form"))
        self.btn_search.setText(_translate("Form", "Browse"))
        self.lineEdit.setText(_translate("Form", "PENGARUH RANDOM UNDER-SAMPLING TERHADAP KLASIFIKASI KECURANGAN TRANSASKI SECARA MOBILE MENGGUNAKAN GRADIENT BOOST TREE "))
        self.label_inputdata.setText(_translate("Form", "Masukkan Dataset"))
        self.btn_clasify.setText(_translate("Form", "Clasify"))
        self.label_cf.setText(_translate("Form", "Confusion Matrix"))
        self.label_TP.setText(_translate("Form", "True Positive"))
        self.label_FP.setText(_translate("Form", "False Positive"))
        self.label_FN.setText(_translate("Form", "False Negative"))
        self.label_TN.setText(_translate("Form", "True Negative"))
        self.label_cf_2.setText(_translate("Form", "Confusion Matrix"))
        self.label_TP_2.setText(_translate("Form", "True Positive"))
        self.label_FP_2.setText(_translate("Form", "False Positive"))
        self.label_FN_2.setText(_translate("Form", "False Negative"))
        self.label_TN_2.setText(_translate("Form", "True Negative"))
        self.label_klasifikasi.setText(_translate("Form", "Klasifikasi"))
        self.label_akurasi.setText(_translate("Form", "Akurasi"))
        self.label_inputdata_2.setText(_translate("Form", "Masukkan Jumlah Pohon"))
        self.checkBox.setText(_translate("Form", "Random Under-sampling"))
        self.btn_open.setText(_translate("Form", "Open"))

scriptDir = dirname(realpath(__file__))
From_Main, _ = loadUiType(join(dirname(__file__), "app.ui"))

class MainWindow(QWidget, From_Main):
    
    def __init__(self):
        super(MainWindow, self).__init__()
        QWidget.__init__(self)
        self.datasetsiap = None
        self.model = model.Model()
        self.ui = Ui_Form()
        self.ui.setupUi(self)
        
        self.ui.btn_search.clicked.connect(self.BrowseFile)
        self.ui.btn_open.clicked.connect(self.OpenFile)
        self.ui.btn_clasify.clicked.connect(self.Klasifikasi)
    
    #method untuk mencari file dan simpan ke sistem
    def BrowseFile(self):
        options = QtWidgets.QFileDialog.Options()
        fileName, _ = QtWidgets.QFileDialog.getOpenFileName(self,"QFileDialog.getOpenFileName()", "","Excel Comma Separated Values File (*.csv)", options=options)
        print(fileName)
        self.model.read_data(fileName)
        self.model.prepare()
        df = self.model.dataset
        self.datasetsiap = pd.DataFrame(data=df)
        print('File siap!!!')
        
    #method untuk open file ke sistem
    def OpenFile(self):
        msg = QtWidgets.QMessageBox()
        msg.setIcon(QtWidgets.QMessageBox.Information)
        msg.setText("Apakah Dataset Sudah dipilih?")
        msg.setInformativeText("Pastikan Dataset Sudah dipilih")
        msg.setWindowTitle("InformationBox")
        msg.setStandardButtons(QtWidgets.QMessageBox.Yes |QtWidgets.QMessageBox.No )
        if msg.exec_()==QtWidgets.QMessageBox.Yes:
            dataset = self.model.get_dataset()
            if(self.ui.checkBox.isChecked()):
                self.datasetsiap = self.model.rus()
            else:
                self.datasetsiap = dataset
                self.ui.lineEdit_jumlahpohon.setDisabled(True)
        else:
            self.message_box_warning()
            
        numColomn = 100
        if numColomn == 0:
            NumRows = len(self.datasetsiap.index)
        else:
            NumRows = numColomn
        self.ui.tableWidget.setColumnCount(len(self.datasetsiap.columns))
        self.ui.tableWidget.setRowCount(NumRows)
        self.ui.tableWidget.setHorizontalHeaderLabels(self.datasetsiap.columns)

        for i in range(NumRows):
            for j in range(len(self.datasetsiap.columns)):
                self.ui.tableWidget.setItem(i, j, QTableWidgetItem(str(self.datasetsiap.iat[i, j])))

        self.ui.tableWidget.resizeColumnsToContents()
        self.ui.tableWidget.resizeRowsToContents()
      
    #method klasifikasi
    def Klasifikasi(self):
        msg = QtWidgets.QMessageBox()
        msg.setIcon(QtWidgets.QMessageBox.Information)
        msg.setText("Apakah Jumlah Pohon Sudah dimasukkan?")
        msg.setInformativeText("*GBT Tanpa RUS Pakai jumlah pohon default")
        msg.setWindowTitle("InformationBox")
        msg.setStandardButtons(QtWidgets.QMessageBox.Yes |QtWidgets.QMessageBox.No )
        if msg.exec_()==QtWidgets.QMessageBox.Yes:
            dataset = self.datasetsiap
            
            num_format = re.compile("^[\-]?[1-9][0-9]*\.?[0-9]+$")
            jmlh_tree = self.ui.lineEdit_jumlahpohon.text()
            isnumber = re.match(num_format, jmlh_tree)
            # if isnumber:
            if(self.ui.checkBox.isChecked()):
                if isnumber:
                    temp = self.model.classifyRUS(dataset, int(jmlh_tree))
                    print(temp)
                    self.ui.lineEdit_akurasi.setText(str(self.model.getAkurasi()))
                    self.ui.lineEdit_TP.setText(str(self.model.getConf_matrix()["TP"]))
                    self.ui.lineEdit_FP.setText(str(self.model.getConf_matrix()["FP"]))
                    self.ui.lineEdit_FN.setText(str(self.model.getConf_matrix()["FN"]))
                    self.ui.lineEdit_TN.setText(str(self.model.getConf_matrix()["TN"]))
                else:
                    self.message_box_error()
            else:
                temp = self.model.classify(dataset)
                print(temp)
                self.ui.lineEdit_akurasi.setText(str(self.model.getAkurasi()))
                self.ui.lineEdit_TP.setText(str(self.model.getConf_matrix()["TP"]))
                self.ui.lineEdit_FP.setText(str(self.model.getConf_matrix()["FP"]))
                self.ui.lineEdit_FN.setText(str(self.model.getConf_matrix()["FN"]))
                self.ui.lineEdit_TN.setText(str(self.model.getConf_matrix()["TN"]))
            # else:
                # self.message_box_error()
        else:
            self.message_box_warning()

    def message_box_warning(self):
        msg=QtWidgets.QMessageBox(QtWidgets.QMessageBox.Warning, "Warning", "Silakan Diperiksa Kembali!")
        retval=msg.exec_()
        print("message Jalan", retval)
        
    def message_box_error(self):
        msg=QtWidgets.QMessageBox(QtWidgets.QMessageBox.Warning, "Error", "Silakan Masukkan dengan Benar!")
        retval=msg.exec_()
        print("message Jalan", retval)
        
#main
if __name__=="__main__":
    app = QApplication(sys.argv)
    sheet = MainWindow()
    sheet.show()
    sys.exit(app.exec_())